### DISCUSSION TITLE
增加llvm依赖链接失败

### DISCUSSION THREAD

**Discussion by plutolove** [xmake-io/xmake#3385] [Q&A]
# build 配置
```lua
add_rules("mode.debug", "mode.release")
add_requires("llvm 14.*")

target("common")
    set_kind("binary")
    add_files("src/jit/*.cpp", "src/*.cpp")
    add_packages("llvm")
```
编译系统环境：clang-15
错需信息：undefined reference to `non-virtual thunk to llvm::RTDyldMemoryManager::anchor()
有很多错误信息，感觉是llvm的目标没有link上，有人遇到过这种情况吗

**Comment by plutolove** [comment]
rence to `vtable for llvm::raw_pwrite_stream'
/home/meng/.xmake/packages/b/binutils/2.38/c69438b2089549c39f3ee587dbbd8cf7/bin/ld: sql_jit.cpp:(.text._ZN6common11JITCompiler7compileERN4llvm6ModuleE[_ZN6common11JITCompiler7compileERN4llvm6ModuleE]+0x7bd): undefined reference to `llvm::raw_ostream::~raw_ostream()'
/home/meng/.xmake/packages/b/binutils/2.38/c69438b2089549c39f3ee587dbbd8cf7/bin/ld: build/.objs/common/linux/x86_64/release/src/jit/sql_jit.cpp.o:(.data.rel.ro._ZTIN6common22JITModuleMemoryManagerE[_ZTIN6common22JITModuleMemoryManagerE]+0x10): undefined reference to `typeinfo for llvm::RTDyldMemoryManager'
/home/meng/.xmake/packages/b/binutils/2.38/c69438b2089549c39f3ee587dbbd8cf7/bin/ld: build/.objs/common/linux/x86_64/release/src/jit/sql_jit.cpp.o:(.data.rel.ro._ZTIN6common17JITSymbolResolverE[_ZTIN6common17JITSymbolResolverE]+0x10): undefined reference to `typeinfo for llvm::LegacyJITSymbolResolver'
/home/meng/.xmake/packages/b/binutils/2.38/c69438b2089549c39f3ee587dbbd8cf7/bin/ld: build/.objs/common/linux/x86_64/release/src/jit/sql_jit.cpp.o:(.data.rel.ro._ZTVN6common22JITModuleMemoryManagerE[_ZTVN6common22JITModuleMemoryManagerE]+0x30): undefined reference to `llvm::RuntimeDyld::MemoryManager::allocateTLSSection(unsigned long, unsigned int, unsigned int, llvm::StringRef)'
/home/meng/.xmake/packages/b/binutils/2.38/c69438b2089549c39f3ee587dbbd8cf7/bin/ld: build/.objs/common/linux/x86_64/release/src/jit/sql_jit.cpp.o:(.data.rel.ro._ZTVN6common22JITModuleMemoryManagerE[_ZTVN6common22JITModuleMemoryManagerE]+0x50): undefined reference to `llvm::RTDyldMemoryManager::registerEHFrames(unsigned char*, unsigned long, unsigned long)'
/home/meng/.xmake/packages/b/binutils/2.38/c69438b2089549c39f3ee587dbbd8cf7/bin/ld: build/.objs/common/linux/x86_64/release/src/jit/sql_jit.cpp.o:(.data.rel.ro._ZTVN6common22JITModuleMemoryManagerE[_ZTVN6common22JITModuleMemoryManagerE]+0x58): undefined reference to `llvm::RTDyldMemoryManager::deregisterEHFrames()'
/home/meng/.xmake/packages/b/binutils/2.38/c69438b2089549c39f3ee587dbbd8cf7/bin/ld: build/.objs/common/linux/x86_64/release/src/jit/sql_jit.cpp.o:(.data.rel.ro._ZTVN6common22JITModuleMemoryManagerE[_ZTVN6common22JITModuleMemoryManagerE]+0x70): undefined reference to `llvm::RTDyldMemoryManager::anchor()'
/home/meng/.xmake/packages/b/binutils/2.38/c69438b2089549c39f3ee587dbbd8cf7/bin/ld: build/.objs/common/linux/x86_64/release/src/jit/sql_jit.cpp.o:(.data.rel.ro._ZTVN6common22JITModuleMemoryManagerE[_ZTVN6common22JITModuleMemoryManagerE]+0xa0): undefined reference to `llvm::RTDyldMemoryManager::getPointerToNamedFunction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)'
/home/meng/.xmake/packages/b/binutils/2.38/c69438b2089549c39f3ee587dbbd8cf7/bin/ld: build/.objs/common/linux/x86_64/release/src/jit/sql_jit.cpp.o:(.data.rel.ro._ZTVN6common22JITModuleMemoryManagerE[_ZTVN6common22JITModuleMemoryManagerE]+0xc8): undefined reference to `llvm::LegacyJITSymbolResolver::lookup(std::set<llvm::StringRef, std::less<llvm::StringRef>, std::allocator<llvm::StringRef> > const&, llvm::unique_function<void (llvm::Expected<std::map<llvm::StringRef, llvm::JITEvaluatedSymbol, std::

**Comment by waruqi** [comment]
llvm 目前默认是 toolchain 包，不提供库，也就是最近才刚刚支持 library 包，但是要自己设置开启，并且目前仅仅提供 几个组件，具体看下

https://github.com/xmake-io/xmake-repo/issues/1802

没有的组件，或者组件内部子库 link 顺序不对的，自己提 pr 过来，改进下包

**Reply by plutolove** [reply]
我和我的cmake项目对比了，llvm的base组件包含我所需要的，那可能是顺序不对，我试试看能不能修复

**Reply by plutolove** [reply]
不行，搞不定，我用动态链接库解决，先用系统自己的llvm依赖吧

**Comment by plutolove** [comment]
https://github.com/xmake-io/xmake-repo/pull/1847 修复了llvm base链接顺序，本地测试通过了

**Reply by plutolove** [reply]
![image](https://user-images.githubusercontent.com/8512477/219934244-e8151d92-830a-4ece-a0b4-f4246a8c8ebd.png)

**Reply by waruqi** [reply]
这只是安装，包本身原本就是 install ok 的，ci 测试过（测试用的符号不多，太不全）。。得 link 到你实际的项目中，如果还是编译 link 通过，说明顺序对了

**Reply by plutolove** [reply]
![image](https://user-images.githubusercontent.com/8512477/220029745-a928f62d-1f18-4776-9ec2-7956952be8f6.png)

**Reply by plutolove** [reply]
![image](https://user-images.githubusercontent.com/8512477/220030014-e62cff3d-fd36-4fdc-a515-fa96370b3c8c.png)

**Reply by plutolove** [reply]
链接的顺序和我这里配置的顺序为啥不一样

**Reply by plutolove** [reply]
![image](https://user-images.githubusercontent.com/8512477/220030331-249b9325-9038-4201-9167-88240e84bbbe.png)

**Reply by plutolove** [reply]
我的编译配置

**Comment by waruqi** [comment]
之前 repo pr merge 后，有没有 `xrepo update-repo` 更新仓库，并且重装 package?

**Reply by plutolove** [reply]
我试试

**Reply by plutolove** [reply]
![image](https://user-images.githubusercontent.com/8512477/220034517-3826f37b-33fe-4586-aff3-bc759d5e16ba.png)
成功了
